### DISCUSSION TITLE
custom fork vladmandic

### DISCUSSION THREAD

**Discussion by motorist828** [lshqqytiger/stable-diffusion-webui-directml#74] [General]
Hi I found this interesting thread https://github.com/vladmandic/automatic
Many are upset about the slowdown in development of AUTOMATIC1111
A person wants to bring people together to work faster and fix problems
Will we ever see directml for this version?

**Comment by OldPixelReaper** [comment]
interesting

**Comment by Aptronymist** [comment]
> Hi I found this interesting thread https://github.com/vladmandic/automatic Many are upset about the slowdown in development of AUTOMATIC1111 A person wants to bring people together to work faster and fix problems Will we ever see directml for this version?

I communicated with vlad last week asking the same thing, he's taking a look at this project and perhaps SHARK, to see what's going to work best for AMD users last I heard, but he said he wanted to not hack it together. I'm looking forward to it, he knows how to get things done quick and he's putting in the effort!

**Reply by Miraihi** [reply]
Tried SHARK just yesterday, and it's surprisingly slower than DirectML, has less features and crashes my drivers as a bonus.

**Reply by motorist828** [reply]
I tried the SHARK, as I understand it, it does not work on Vega video cards :(

**Reply by Omniphantasm** [reply]
> Tried SHARK just yesterday, and it's surprisingly slower than DirectML, has less features and crashes my drivers as a bonus.

I've never even been able to get it to create a single image.

POS Crashes me back to prompt as soon as it makes the VAE or whatever. Can't find a single other person with the issue on a 7900XTX so I just gave up and have decided its a bad program as is logical.

**Comment by Lumpsack** [comment]
Also throw NCNN into the ring for AMD support, I don't truly know how it compares, but ChainNer uses it and its really fast compared with the CPU

**Comment by Lumpsack** [comment]
NM, I found the thread and put the question there - thanks for highlighting it.

**Comment by lshqqytiger** [comment]
I wanted to create a new branch for this, but it seemed difficult. (due to `checkout master`)
I created [a new repository](https://github.com/lshqqytiger/vladmandic-automatic-directml) and you can try it.

I recommend not to use `webui.bat` or `webui.sh`.
Activate your virtual env using python venv or anaconda and start webui with `python launch.py`.

New installation guide:
(TBA)

Memory optimization for AMDGPUs is disabled by default. You can enable it on `Settings > Stable diffusion > Enable memory optimization for AMDGPUs. Only works for DirectML backend.`

If black images are generated, enable no-half on `Settings > CUDA Settings > Use full precision for model (--no-half)` and `Use full precision for VAE (--no-half-vae)`.

You can apply sub quadratic attention (`--opt-sub-quad-attention`) on `Settings > Stable diffusion > Cross-attention optimization method`.

**Reply by motorist828** [reply]
It doesn't seem to work, but I'm not sure if I ran it correctly
Maybe add "launch.bat" for inexperienced users
![Screenshot_83](https://user-images.githubusercontent.com/40582999/233839288-fa512b9e-bda3-4e95-b050-fd8a82033ed3.png)
![Screenshot_84](https://user-images.githubusercontent.com/40582999/233839294-1f1d614b-7fb2-4342-8e2c-42ac990cb690.png)

When I run webui.bat I get this error
![Screenshot_85](https://user-images.githubusercontent.com/40582999/233839950-5e999276-9c94-4573-98e3-c2c522875cfb.png)

Added my model, got this
![Screenshot_86](https://user-images.githubusercontent.com/40582999/233840245-b0def563-26a1-41aa-b0e4-3c14666cac8e.png)

**Reply by Miraihi** [reply]
I launched via launch.py like you recommended, but still get
```
WARNING  Package wrong version: torch 2.0.0 required 1.13.1
16:38:07-831716 ERROR    Error running pip with args: install --upgrade torch==1.13.1 torchvision==0.14.1
                         torch-directml --force-reinstall --ignore-installed
```

When trying to generate literally anything I get out of memory error like
`RuntimeError: Could not allocate tensor with 1800000000 bytes. There is not enough GPU video memory available!`
Nonetheless, the performance is pretty similar to your original implementation, maybe slightly faster. About 1.13 s/it with DPM++ 2M Karras for 512x512.

**Reply by Omniphantasm** [reply]
> Memory optimization for AMDGPUs is disabled by default. You can enable it on `Settings > Stable diffusion > Enable memory optimization for AMDGPUs. Only works for DirectML backend.`

Probably just being a smooth brain but I ain't seeing that option on there my man. I even hit show all pages and ctrl-f'd for AMD

**Reply by motorist828** [reply]
For two hours, I still could not install it correctly
It started once, but when trying to generate the image, a nohalf error occurred, after that it didn’t start again

**Reply by lshqqytiger** [reply]
Setup does not smoothly work because DirectML does not support torch 2.0.0. Remove venv and retry. If the same issue remains, please attach setup.log.

Performance is similar to `stable-diffusion-webui-directml`, which is already optimized enough.

**Reply by lshqqytiger** [reply]
> Probably just being a smooth brain but I ain't seeing that option on there my man. I even hit show all pages and ctrl-f'd for AMD

Check whether you cloned correct repository. You should clone https://github.com/lshqqytiger/vladmandic-automatic-directml.

**Reply by motorist828** [reply]
>Remove venv and retry.

How to properly activate venv?
The commands I found didn't work for me

**Reply by Omniphantasm** [reply]
> > Probably just being a smooth brain but I ain't seeing that option on there my man. I even hit show all pages and ctrl-f'd for AMD
> 
> Check whether you cloned correct repository. You should clone https://github.com/lshqqytiger/vladmandic-automatic-directml.

Ah there's where I went smoothbrain, I followed the instructions on the page you linked and failed to notice the difference of

git clone https://github.com/vladmandic/automatic

instead of 

git clone https://github.com/lshqqytiger/vladmandic-automatic-directml

Thank you 😁

Edit: Looks like I'm getting the same name '_C' is not defined. That Motorist is, completely fresh clone, it launches server just fine. Shut down server, restart it with webui.bat and it spits out that _C error. Something is messing up on second runs.

Edit Edit: Motorist helped me get it set up and now its working amazingly.

**Reply by motorist828** [reply]
This worked for me
I had a problem with python that didn't affect other versions of sd
From the first tests I see a speed increase of 33s versus 39s (26 steps, DPM ++ 2M Karras, 640x640, vega56)
This is good

**Reply by Omniphantasm** [reply]
Do share your exact steps for getting it working cuz I'm getting the same _C crap you were and I'm even dumber about setting up a venv, everything I'm finding is assuming I'm the one writing a program and want that program to use a venv.

**Reply by motorist828** [reply]
> Do share your exact steps 

Delete the venv folder
run CMD in vladmandic-automatic-directml directory
`python -m venv venv`
`venv\Scripts\activate `
`python launch.py`

After that I needed to reinstall torch

`pip install torch==1.13.1 torch-directml --force-reinstall --ignore-installed`
launch, it works 
`python launch.py --medvram`

**Reply by Omniphantasm** [reply]
Nice, that did it for me. Thanks

Now is there a particular way I need to start and/or update this folder in the future so I don't horribly break everything again? 😆

Holy hell, it's so much faster now. 512x512 set of 8 only took 30 seconds, it's actually faster than my old A1111 setup was going before I borked it the other night.

**Reply by motorist828** [reply]
Looks like everything is broken today.
used directml-fix.bat before launch
![Screenshot_90](https://user-images.githubusercontent.com/40582999/233990962-830661a8-741e-4b2f-8124-dc06f3c13f92.png)
![Screenshot_89](https://user-images.githubusercontent.com/40582999/233990982-991ba698-b9b7-4d9c-bc6e-2c202e66d717.png)

**Reply by lshqqytiger** [reply]
It worked well before I push, but it was broken for me, too. I'm finding way to make it work again.

**Reply by lshqqytiger** [reply]
Now, setup seems to be work well.
But there is a same issue with https://github.com/vladmandic/automatic/issues/448

**Comment by Yodrik** [comment]
I tried this new fork. *1

I think I goofed up the venv thing, but I ran into "RuntimeError: ... There is not enough GPU video memory available!" at first.
Subsequently I ran on --medvram which works to generate 512x512, but I was unable to generate 600x800 regardless of the other options enabled.

I took 7 samples while trying to adjust settings to get to 600x800,
(RX 6600) 512x512 image times @ Steps: 10, Sampler: DPM++ 2M Karras 

| | vladmandic-automatic-directml | stable-diffusion-webui-directml |
| ------------- | ------------- | ------------- |
Time taken | 48.85s | - |
| - | 45.56s | 
| - | 40.15s | 
| - | 45.28s | 
| - | 45.07s |
| - | 46.20s |
| - | 44.88s |
| Avg. | ~45s | ~16s |

Unfortunately right now it almost feels like the --lowvram speed on this fork.

*1 stable-diffusion-webui-directml borked (steps run, but no image generated) after the testing and I had to reinstall python + reinstall stable-diffusion-webui-directml to get it to work again. 
I mean to do more adjusting and take image time data in this fork before playing around with the vladmandic fork again.

**Comment by vladmandic** [comment]
i love the effort here! @lshqqytiger any chance you'd be interested in contributing to main repo directly?
i'm more than open to merge anything thats needed, just need to make it conditional - so it works for standard as well as directml use-case.

**Reply by lshqqytiger** [reply]
Yes I am. But this repository has some codes that are not suitable for merging into the main repository. I need to remove them to merge this. I agree to offer this as an option and would like to help it.

**Reply by vladmandic** [reply]
perfect! if you need anything, let me know.

**Reply by lshqqytiger** [reply]
Can I access a branch for directml implementation? I will compare this repo with yours, and then I will push the necessary parts. I'll work only on that branch. You can merge it into master whenever you want.

**Reply by vladmandic** [reply]
as far as i know, github doesn't have per-branch user-rights, but i don't mind creating a branch and making you a collaborator in general - with understanding that you won't push to master blindly :)

**Reply by lshqqytiger** [reply]
okay thank you :)

**Reply by vladmandic** [reply]
done. and i've created a branch for you: <https://github.com/vladmandic/automatic/tree/directml>

**Reply by vladmandic** [reply]
Dicord: vlado#8083

**Comment by lshqqytiger** [comment]
This moved to https://github.com/vladmandic/automatic/tree/directml.
I didn't test this branch yet, and I just moved some parts with comments.
I'll test it tomorrow.

**Reply by motorist828** [reply]
Tested a little
Installed without problems, everything is fine
But slow speed, this is probably due to the lack of "Enable memory optimization for AMDGPUs"

**Reply by Omniphantasm** [reply]
You installed the master, you need the directml branch.

`git clone -b directml https://github.com/vladmandic/automatic`

**Reply by motorist828** [reply]
> `git clone -b directml https://github.com/vladmandic/automatic`

i know i did it
Maybe an update will fix it
a lot of commits came out today

**Reply by Omniphantasm** [reply]
Ah, I'll try them out. I'm on the old one from before he changed over to that repo and I keep having issues where it runs fine and fast. Then I shut down server come back later and start it up and every generation just bitches at me 

> RuntimeError: The GPU device instance has been suspended. Use GetDeviceRemovedReason to determine the appropriate
> action.

Hopefully theres been a commit that fixes it cuz damned if I know whats causing it. I just reboot and it goes away it seems.

Edit: Yup, I see what you were saying. The AMD option is indeed missing for me as well. My bad for assuming 😁

**Reply by Cwieku666** [reply]
I checked a few things, for example the built in token merger works and I was able to do batch of 6 images with a resolution of 512x512 with 4gb of vram. Unfortunately best extension for low vram cards tiled vae doesn't work and throws this error, AttributeError: 'Namespace' object has no attribute 'sub_quad_q_chunk_size' https://i.imgur.com/xe5JlRn.png

**Reply by lshqqytiger** [reply]
Did you use hires fix? When `Apply only to high-res fix pass. Disabling can yield a ~20-35% speedup on contemporary resolutions.` is checked, token merging works only for hires fix and DirectML does not have operators token merging requires.

`AttributeError: 'Namespace' object has no attribute 'sub_quad_q_chunk_size'`
This error is occurred because cmd_opts is separated into cmd_opts/opts in `vladmandic/automatic`.
cmd_opts does not have `sub_quad_q_chunk_size`, opts does.

**Reply by Miraihi** [reply]
I don't know what I'm doing wrong, but I can't generate anything. Getting the out of memory error each time.

**Reply by vladmandic** [reply]
> Unfortunately best extension for low vram cards tiled vae doesn't work and throws this error

Which extension is that? I can add compatibility option like i've done for plenty other extensions.

**Reply by lshqqytiger** [reply]
> > Unfortunately best extension for low vram cards tiled vae doesn't work and throws this error
> 
> Which extension is that? I can add compatibility option like i've done for plenty other extensions.

It is multidiffusion. Only when sub-quadratic attention selected.

**Reply by vladmandic** [reply]
> It is multidiffusion. Only when sub-quadratic attention selected.

link?

**Reply by lshqqytiger** [reply]
Builtin extension. https://github.com/pkuliyi2015/multidiffusion-upscaler-for-automatic1111

Right here: https://github.com/pkuliyi2015/multidiffusion-upscaler-for-automatic1111/blob/main/tile_utils/attn.py#L185-L185

**Reply by lshqqytiger** [reply]
> I don't know what I'm doing wrong, but I can't generate anything. Getting the out of memory error each time.

Very strange. How about this commit?: `8ea2fe15e755a7c92c8735f2eedde7092fd4f1aa`

**Reply by vladmandic** [reply]
fixed.

**Reply by Miraihi** [reply]
Hey, now it works completely fine. Thanks for the speedy solution!

**Reply by Miraihi** [reply]
What's cool is that Approx NN live preview works. Didn't work in automatic111-directml.

**Reply by motorist828** [reply]
Tested with latest updates
Everything seems to work, though there is no increase in speed of 16% that was in the very first release
Hmm maybe i did something wrong then

**Reply by Miraihi** [reply]
All of the sudden I'm getting this error when loading the webUI
Commit `9a224a5cb83cbcf75ab0108c03196b8519d29ab6`
I did try to delete `venv`, didn't work.
```
Rollback VAE functionality requires Torch 2.1 or higher

┌───────────────────────────────────────── Traceback (most recent call last) ─────────────────────────────────────────┐
│ E:\AI\vladmandic-directml\launch.py:100 in <module>                                                                 │
│                                                                                                                     │
│    99     import webui                                                                                              │
│ > 100     webui.webui()                                                                                             │
│   101                                                                                                               │
│                                                                                                                     │
│ E:\AI\vladmandic-directml\webui.py:216 in webui                                                                     │
│                                                                                                                     │
│   215 def webui():                                                                                                  │
│ > 216     start_ui()                                                                                                │
│   217     load_model()                                                                                              │
│                                                                                                                     │
│ E:\AI\vladmandic-directml\webui.py:169 in start_ui                                                                  │
│                                                                                                                     │
│   168     create_paths(opts)                                                                                        │
│ > 169     initialize()                                                                                              │
│   170     if shared.opts.clean_temp_dir_at_start:                                                                   │
│                                                                                                                     │
│ E:\AI\vladmandic-directml\webui.py:78 in initialize                                                                 │
│                                                                                                                     │
│    77 def initialize():                                                                                             │
│ >  78     check_rollback_vae()                                                                                      │
│    79                                                                                                               │
│                                                                                                                     │
│ E:\AI\vladmandic-directml\webui.py:72 in check_rollback_vae                                                         │
│                                                                                                                     │
│    71             shared.cmd_opts.rollback_vae = False                                                              │
│ >  72         if 0 < torch.cuda.get_device_capability()[0] < 8:                                                     │
│    73             print('Rollback VAE functionality device capabilities not met')                                   │
│                                                                                                                     │
│ E:\AI\vladmandic-directml\venv\lib\site-packages\torch\cuda\__init__.py:357 in get_device_capability                │
│                                                                                                                     │
│   356     """                                                                                                       │
│ > 357     prop = get_device_properties(device)                                                                      │
│   358     return prop.major, prop.minor                                                                             │
│                                                                                                                     │
│ E:\AI\vladmandic-directml\venv\lib\site-packages\torch\cuda\__init__.py:371 in get_device_properties                │
│                                                                                                                     │
│   370     """                                                                                                       │
│ > 371     _lazy_init()  # will define _get_device_properties                                                        │
│   372     device = _get_device_index(device, optional=True)                                                         │
│                                                                                                                     │
│ E:\AI\vladmandic-directml\venv\lib\site-packages\torch\cuda\__init__.py:221 in _lazy_init                           │
│                                                                                                                     │
│   220         if not hasattr(torch._C, '_cuda_getDeviceCount'):                                                     │
│ > 221             raise AssertionError("Torch not compiled with CUDA enabled")                                      │
│   222         if _cudart is None:                                                                                   │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
AssertionError: Torch not compiled with CUDA enabled
Press any key to continue . . .
```

**Reply by lshqqytiger** [reply]
`rollback_vae` is disabled by default. Open `config.json` and edit `rollback_vae` to false.

**Reply by Miraihi** [reply]
Right, I've just realized it's been caused by some config changes. I think it won't hurt to hide that option.

**Reply by Miraihi** [reply]
If after reloading the UI I try to get my last probpt back via the dedicated button, I get the error

```
Traceback (most recent call last):
  File "E:\AI\vladmandic-directml\venv\lib\site-packages\gradio\routes.py", line 394, in run_predict
    output = await app.get_blocks().process_api(
  File "E:\AI\vladmandic-directml\venv\lib\site-packages\gradio\blocks.py", line 1078, in process_api
    data = self.postprocess_data(fn_index, result["prediction"], state)
  File "E:\AI\vladmandic-directml\venv\lib\site-packages\gradio\blocks.py", line 994, in postprocess_data
    block = self.blocks[output_id]
KeyError: 331
```

**Reply by Miraihi** [reply]
Unfortunately "Tiled VAE" (Mutidiffusion) still produces grey tiles if the decoder tile size is insufficient. No improvements here.

**Reply by Miraihi** [reply]
Big improvement over automatic-directml: In img2img you need less steps the less Denoising strength you have set. I think I'll be using Krita plugin with Vladmandic version now.

**Reply by Miraihi** [reply]
Possible improvement: [Ratio helper](https://github.com/thomasasfk/sd-webui-aspect-ratio-helper) to be a part of the default package. Really helps with the latent upscaling task.

**Reply by Yodrik** [reply]
I did not finish all the sampling I wanted to, but 2 observations and 2 questions.

1. Though a lot of the performance adjustments are in UI now, it seems a proper close/restart is necessary to see the changes in effect (from testing and the System Info). I believe vladmandic is already aware of issues surrounding "restart" from the UI.
2. I observed better speed using -upcast sampling in place of -no-half. I will gather time data. I never used this previously.
3. To confirm, if we want to enable --medvram/lowvram it is still a command arg? Oddly the upcast speed observation and running 512x512 without much of an issue were all with --medvram not active. 
4. the UI has the default sub-quad qchunk/kvchunk/vram% as 512/512/80%. The [automatic1111 wiki](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Command-Line-Arguments-And-Settings) has defaults of 1024/none (implied 1024?)/none (implied100%?). Can we assume the chunk default has always been as the wiki says and is different from this fork? This is merely a clarification and not a priority question.

**Reply by motorist828** [reply]
> 2\. I observed better speed using -upcast

what gpu do you have?
i get the opposite result using vega56
for me it's about 16% slower

**Reply by Yodrik** [reply]
> > 2. I observed better speed using -upcast
> 
> what gpu do you have? i get the opposite result using vega56 for me it's about 16% slower

RX 6600 8GB
~~I was hesitant to value any data from that observation, because my reading up on upcast sampling is that it was mainly beneficial for older NVIDIAs alongside the talk about RDNA2 cards may or may not needing --precision-full --no-half in the first place.~~ I need to spend time with that stuff removed to eliminate other possibilities. 

edit: I found you are correct, but I can't replicate the same setting in new fork right now.

**Reply by Yodrik** [reply]
I confirmed in my auto1111 I was only seeing the difference between  [--precision-full --no-half ] and [--precision-full --upcast-sampling]. I confirmed [--precision-full] is quicker than [--precision-full --upcast-sampling] and [--precision-full --no-half ] which aligns with motorist828.

But trying to replicate that same setting in the new fork gives me a new error, "RuntimeError: mat1 and mat2 must have the same dtype" which does not happen in auto1111. [--precision-full --upcast-sampling] and [--precision-full --no-half ] seem to be working though.
I'm guessing this error is [https://github.com/vladmandic/automatic/issues/470#issue-1682116401](url) . 
I will try different checkpoints tomorrow.

**Reply by Miraihi** [reply]
@Yodrik upcast sampling is not about speed, it's more about vram conservation. So, it means "going out of memory less".

**Reply by vladmandic** [reply]
upscast sampling is less aggressive version of no-half-vae - it helps in most of cases where no-half-vae is needed, but with less performance penalties.

**Reply by Cwieku666** [reply]
I don't know why there is such a big difference between Auto1111 and Vlad but with 4gb vram card on auto even 512x512 with hires fix to 640x640 is too much, but with Vlad fork I can generate 512x512 with hires fix 1.9 to 972x972. 
My cuda settings: https://i.imgur.com/mOcyaSw.png
and successfully generated 972x972: https://i.imgur.com/ziirA6B.png or batch of 2 images upscaled to 896x896 https://i.imgur.com/h7h4pnn.png

**Reply by motorist828** [reply]
>  4gb vram card 

What GPU are you using?
With your settings I got a size increase too
Can generate at least 1520x1520 but slow 48.26s/it (vega56)
this saves memory well but slows down speed for me by 40-50%

**Reply by Cwieku666** [reply]
> > 4gb vram card
> 
> What GPU are you using? With your settings I got a size increase too Can generate at least 1520x1520 but slow 48.26s/it (vega56) this saves memory well but slows down speed for me by 40-50%

RX 470, so far i haven't compared the speed between different settings, i was checking how far i can go on square ratio.
Unfortunately tiled vae stopped working for me with these settings.

**Reply by Miraihi** [reply]
If your GPU doesn't support fp-16 anyway, then running it with upcast-sampling or autocast instead of no-half/full precision is going to just "distract" it from the task (It will be trying to use half precision calculations, but never succeed), thus the slowdown. In addition, it's likely to break the inpainting.

**Reply by Cwieku666** [reply]
both test with same settings and --lowvram, Euler A 20 steps 512x512 with hires x1.9 to 972x972
upcast sampling, without --no-half/--no-half-vae - successfully generated Time taken: 5m 41.50s
with --no-half and --no-half-vae, with turned off upcast sampling - out of memory with Time taken: 6m 32.45s
just don't know why with upcast sampling tiled vae doesn't work and with --no-half it works

Edit: with upcast samling for the first time I was able to generate something with --medvram, it was not possible on auto at all and so far max resolution with medvram for me was 512x512 with 1.5 upscale to 768x768 (Time taken: 2m 16.98s)

anoter thing is the difference between lowvram and medvram, even on 20 steps at 512x512 without hires is big
--lowvram - 1m 7.94s
--medvram - 33.20s
and why with lowvram quality is so bad? lowvram-> https://i.imgur.com/NyxIfHY.jpeg and medvram->https://i.imgur.com/hKFGkw3.jpeg

**Reply by Yodrik** [reply]
> Edit: with upcast samling for the first time I was able to generate something with --medvram, it was not possible on auto at all and so far max resolution with medvram for me was 512x512 with 1.5 upscale to 768x768 (Time taken: 2m 16.98s)

You might want to check on what your systeminfo tab says about Torch. --medvram --upcast-sampling won't change what you have set in the settings for Precision Autocast or Full (the radio buttons). So in system info it should show "**torch:1.13.1+cpu Autocast  half**". 

If you switch it to full, after restarting you will see "**torch:1.13.1+cpu full half**" which would bring you equivalent to --medvram --precision full --upcast-sampling. From there you can switch upcast off and see how that is, or switch it off and turn on -nohalf ( which gives you **torch:1.13.1+cpu Full  nohalf** ). That is the only set which I found the benchmark test to complete so at the minimum 512x512 squares in low and high batches are producible.

Regarding lowvram and medvram you may wish to read this thread, #84 .

**Reply by Miraihi** [reply]
@Cwieku666 My advice: Stop using hirezfix and begin using [Ultimate SD upscale script](https://github.com/Coyote-A/ultimate-upscale-for-automatic1111). Hirezfix is way too memory-hungry on directml. Multidiffusion is also great, but produces grey tiles if you can't affort to raise the Vae decoder tile size. I have some of your args questions answered in [my article](https://github.com/lshqqytiger/stable-diffusion-webui-directml/discussions/84).

**Reply by vladmandic** [reply]
@Miraihi nice article - thanks for sharing

**Reply by diffusersd** [reply]
hires fix seems to be unusable at the moment(with 12gb vram, 6700 xt) and there are few issues with token reduction as well. looking forward for optimizations! I was able to use hires fix in original repo without any issues.

**Reply by Cwieku666** [reply]
Yesterday lshqqytiger added support for torch 2.0 and torch-directml 0.2.0.dev230426 so we have access to SDP optimization. Anyone tested Scaled-Dot-Product optimization with better than my 470 4gb card?
At least for me is slower than sub-quadratic and I get an out of memory error faster
30 steps DMP++ 2m Karras 512x768 with noise offset lora
--lowvram with upcast sampling
sub-quadratic - 2m 9.73s
Scaled-Dot-Product - 2m 22.35s

--lowvram (--no-half), (--no-half-vae), upcast sampling
sub-quadratic - 3m 9.25s, oom on the second try
Scaled-Dot-Product - instant out of memory

--lowvram , (--no-half-vae), upcast sampling
sub-quadratic - 2m 7.54s
Scaled-Dot-Product -2m 8.70s

--lowvram (--no-half), (--no-half-vae)
sub-quadratic - 2m 36.12s
Scaled-Dot-Product -  2m 53.69s

--medvram with upcast sampling
sub-quadratic - 1m 19.58s
Scaled-Dot-Product - instant out of memory

--medvram (--no-half-vae), upcast sampling
sub-quadratic - 1m 18.17s
Scaled-Dot-Product - instant out of memory

--medvram (--no-half), (--no-half-vae)
sub-quadratic - instant out of memory
Scaled-Dot-Product - instant out of memory

--medvram (--no-half), (--no-half-vae), upcast sampling
sub-quadratic - instant out of memory
Scaled-Dot-Product - instant out of memory

**Reply by lshqqytiger** [reply]
SDP attention was more inefficient than sub-quadratic attention for me, too.
It is so disappointing.

**Reply by motorist828** [reply]
I don't see speed changes after upgrading to torch 2.0 (--medvram--no-half --no-half-vae)
It also does not change if I enable the "SDP disable memory attention" option.
Scaled-Dot-Product - instant out of memory (640х640 + medvram)

**Reply by Miraihi** [reply]
Well, looks like torch 2.0 is not really faster even for the baseline Nvidia CUDA version. Not even talking about DIrectML what's another can of worms.

**Reply by vladmandic** [reply]
@Miraihi Torch is not magic, its a framework and it uses CUDA/cuDNN/ROCm for actual math ops. So newer torch enables new features, but any performance gains come from underlying libs that implements ops that are defined in torch. I need new torch as i want to start using some of the new features.

**Reply by motorist828** [reply]
@Miraihi torch 2.0 gives me a very good boost on ROCm linux
So this is more of a problem with DIrectML and other libraries.

**Reply by Miraihi** [reply]
I tried running SD on Linux ROCm, but looks like RX 580 doesn't support it. What a shame.

**Reply by Yodrik** [reply]
A day or two with the merged commit and I sort of settled on the last combo here. SDP was really off and on, either OOM or success. In the benchmark it didn't work out at all though. I understand from the code the benchmark test is 512x512 batches, so it serves as a simple test after install. High res is another story still.

(RX6600 8GB)
| Precision  | -nohalf/-upcast  | MO  | -arg  | benchmark score  |
| ------------ | ------------ | ------------ | ------------ | ------------ |
| Full  | -nohalf  | SubQ  |   | 0.22 / error / error  |
| Full  | -nohalf  | SubQ  | medvram  | 0.27 / 0.27 / 0.24  |
| Full  | -nohalf  | SDP  | medvram  | error / error / error  |
| Full  | -nohalf  | doggettx  | medvram  | 0.3 / 0.27 / 0.25  |
| Full  | -nohalf  | invokeai  | medvram  | error / error / error  |
| Full  | -upcast  | SubQ  | medvram  | 0.46 / 0.4 / 0.37  |

**Reply by Cwieku666** [reply]
with normal benchmark I get 0.63 / 0.53 / 0.55 with rx 470 4gb (Full + upcast + sub-quadratic with medvram) for example somebody with rtx 3060 scored 6.67 / 7.68 / 8.48

quick question how I can change branch to main for new updates when I installed it with git clone -b directml https://github.com/vladmandic/automatic

**Reply by lshqqytiger** [reply]
```
$ git checkout main
```

**Comment by vladmandic** [comment]
you've enabled setting in cuda params that is not compatibile (rollback vae).
i just did the update to check for cuda available before even checking that setting.
update repo and after ui starts, go to settings and disable it.

**Comment by OldPixelReaper** [comment]
Hi, if I want to test it under AMD I need to run 
git clone https://github.com/vladmandic/automatic or
git clone https://github.com/vladmandic/automatic/tree/directml
? Thanks :-)

**Reply by Yodrik** [reply]
As of the latest commits DirectML support is now part of the master branch, so [https://github.com/vladmandic/automatic](url)

*DirectML to use AMD in Windows. This is alongside the existing ROCm to use AMD in Linux already available.

**Reply by OldPixelReaper** [reply]
thx :-)

**Reply by Lumpsack** [reply]
Great piece of news, thanks @Yodrik  - I've just installed main and its working well under AMD using directml natively.
haha, now just to get ROCm into Windows!

**Comment by OldPixelReaper** [comment]
For anyone with AMD on Windows, does the fork still work with the latest updates? For me it seems to run only on the CPU after the git pull. Even a fresh install doesn't do anything. From 1.2 it/s to 24 it/s :(

AMD RX 5700 XT, Win 11

**Reply by vladmandic** [reply]
DirectML was never intended to be enabled by default, you nedd to run with `webui --use-directml`

**Reply by OldPixelReaper** [reply]
I understand that already and that's what I'm trying to do^^

C:\Users\Test\Diff\SDNextUI\automatic\webui.bat --use-directml --medvram

When I do it this way, it shows up:
--
  Using DirectML Backend
00:57:17-410500 INFO     Torch 2.0.0+cpu
00:57:17-410500 WARNING  Torch reports CUDA not available
00:57:17-410500 INFO     Verifying requirements
00:57:17-454628 INFO     No changes detected: Quick launch active
00:57:17-456639 INFO     Running extension preloading
00:57:17-458648 INFO     Server arguments: ['--use-directml', '--medvram']
---


But it still runs as slow. Do I possibly need to put this in a completely different place? Because I can't append "webui --use-directml" there, then I get an (webui) error.

**Reply by vladmandic** [reply]
that startup sequence looks ok. i need to dig deeper into this.

**Reply by OldPixelReaper** [reply]
Maybe this will help you a little bit. Fresh installation, only a few extensions turned off and so, that's all I did. As I said, about 2 weeks ago it ran great. With the fork of lshqqytiger it also runs. I just wait and see now^^

[config.txt](https://github.com/lshqqytiger/stable-diffusion-webui-directml/files/11468737/config.txt)
[setup.log](https://github.com/lshqqytiger/stable-diffusion-webui-directml/files/11468738/setup.log)
[console-startup-debug.txt](https://github.com/lshqqytiger/stable-diffusion-webui-directml/files/11468739/console-startup-debug.txt)
[DxDiag.txt](https://github.com/lshqqytiger/stable-diffusion-webui-directml/files/11468740/DxDiag.txt)

**Reply by howDoILearnPython** [reply]
Wow... let me try that fork too.

**Reply by vladmandic** [reply]
He did all the work to bring DirectML support here.
I'll check the arg parsing, best guess is thar it somehow it skips it even if it's specified and dected by launcher.

**Reply by howDoILearnPython** [reply]
tried fresh install just 30 mins ago, with --use-directml --medvram
there are info said it's using directML backend.   
However, i tried txt2img, according to the speed and task manager, it's apparently still using CPU rather than GPU.  
let me try again and report to here later.

**Reply by howDoILearnPython** [reply]
tried again and confirm all setting is correct.  
inference will cause CPU 100% full load with system memory being used, 20s/it.
FYI, for lshqqytiger fork, GPU will be 100% full load with VRAM being used, 5s/it for my vega 8, 3200G APU.

not sure if this help, following the my console output.

C:\vladmandic>webui --use-directml --medvram --listen
Using VENV: C:\vladmandic\venv
12:24:35-124855 INFO     Starting SD.Next
12:24:35-124855 INFO     Python 3.10.9 on Windows
12:24:35-218782 INFO     Version: a2485cf7 Fri May 12 21:12:24 2023 -0400
12:24:35-624871 INFO     Setting environment tuning
12:24:35-624871 INFO     Using DirectML Backend
12:24:38-171820 INFO     Torch 2.0.0+cpu
12:24:38-171820 WARNING  Torch reports CUDA not available
12:24:38-187447 INFO     Verifying requirements
12:24:38-265578 INFO     No changes detected: Quick launch active
12:24:38-281203 INFO     Running extension preloading
12:24:38-281203 INFO     Server arguments: ['--use-directml', '--medvram', '--listen']
No module 'xformers'. Proceeding without it.
12:24:47-361189 INFO     Libraries loaded
12:24:47-376815 INFO     Available models: C:\vladmandic\models\Stable-diffusion 1
ControlNet v1.1.152
ControlNet v1.1.152
Image Browser: ImageReward is not installed, cannot be used.
12:24:53-049091 INFO     Loading UI theme: black-orange
Running on local URL:  http://0.0.0.0:7860

To create a public link, set `share=True` in `launch()`.
12:25:00-893391 INFO     Local URL: http://localhost:7860/
12:25:00-893391 INFO     API Docs: http://localhost:7860/docs
12:25:00-893391 INFO     Initializing middleware
12:25:01-830958 WARNING  Torch FP16 test failed: Forcing FP32 operations
12:25:01-846584 INFO     Setting Torch parameters: dtype=torch.float32 vae=torch.float32
                         unet=torch.float32
Loading weights: C:\vladmandic\models\Stable-diffusion\beautypromix_v1.safetensors ------ 0.0/3… -:--:--
                                                                                          GB
LatentDiffusion: Running in eps-prediction mode
DiffusionWrapper has 859.52 M params.
12:25:02-893533 INFO     Model created from config: C:\vladmandic\configs\v1-inference.yaml
12:25:03-456076 WARNING  Scaled dot product cross attention is not available on CPU
12:25:03-456076 INFO     Embeddings loaded: 0 []
12:25:03-471698 INFO     Model loaded in 2.2s (load=1.2s create=0.5s apply=0.5s)
12:25:03-706089 INFO     Model load finished: {'ram': {'used': 6.71, 'total': 23.94}}
12:25:04-081117 INFO     Startup time: 25.8s (torch=5.6s gradio=2.4s libraries=1.1s codeformer=0.2s
                         scripts=5.1s opts onchange=0.3s ui=3.5s start=4.3s scripts
                         app_started_callback=0.1s checkpoint=3.1s)
12:25:04-096742 INFO     Memory used: 4.72 total: 23.94

**Comment by OldPixelReaper** [comment]
Seems to be running again, big thanks!
Starting SD.Next
16:11:40-458829 INFO     Python 3.10.11 on Windows
16:11:40-521334 INFO     Version: 554f2629 Wed May 17 08:46:57 2023 -0400
16:11:40-927610 INFO     Setting environment tuning
16:11:40-927610 INFO     Using DirectML Backend
16:11:42-303086 INFO     Torch 2.0.0+cpu
16:11:42-318712 INFO     Torch backend: DirectML (torch-directml 0.2.0.dev230426)
16:11:42-396842 INFO     Torch detected GPU: AMD Radeon RX 5700 XT
